Use StringParsingBuffer in HTMLFastPathParser

cdumez opened the pull request:
#### 7b649bcd89479eba4c15e3a2cd66fe6edc9fa029
<pre>
Use StringParsingBuffer in HTMLFastPathParser
<a href="https://bugs.webkit.org/show_bug.cgi?id=252407">https://bugs.webkit.org/show_bug.cgi?id=252407</a>

Reviewed by Darin Adler.

Use StringParsingBuffer in HTMLFastPathParser. It makes the code clearer and
look safer. This is performance neutral on Speedometer.

* Source/WTF/wtf/text/StringParsingBuffer.h:
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::isQuoteCharacter):
(WebCore::isValidUnquotedAttributeValueChar):
(WebCore::HTMLFastPathParser::HTMLFastPathParser):
(WebCore::HTMLFastPathParser::parseCompleteInput):
(WebCore::HTMLFastPathParser::scanText):
(WebCore::HTMLFastPathParser::scanEscapedText):
(WebCore::HTMLFastPathParser::scanTagName):
(WebCore::HTMLFastPathParser::scanAttributeName):
(WebCore::HTMLFastPathParser::scanAttributeValue):
(WebCore::HTMLFastPathParser::scanEscapedAttributeValue):
(WebCore::HTMLFastPathParser::scanHTMLCharacterReference):
(WebCore::HTMLFastPathParser::parseChildren):
(WebCore::HTMLFastPathParser::parseAttributes):
(WebCore::HTMLFastPathParser::parseContainerElement):
(WebCore::HTMLFastPathParser::isValidUnquotedAttributeValueChar): Deleted.
(WebCore::HTMLFastPathParser::skipWhitespace): Deleted.
(WebCore::HTMLFastPathParser::peekNext): Deleted.
(WebCore::HTMLFastPathParser::consumeNext): Deleted.

Canonical link: <a href="https://commits.webkit.org/260417@main">https://commits.webkit.org/260417@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9accd2ba5b4515a167ce6427251b235c0365956e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108251 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17341 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41116 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117370 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112138 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/18886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8620 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100465 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114020 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/18886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97310 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28952 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/97450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10186 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30297 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/96801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8297 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10922 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/96801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16335 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49893 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105838 "Built successfully") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/7205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12508 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/26201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3917 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -495,12 +500,6 @@ class HTMLFastPathParser {
         return c == ' ' || c == '>' || isHTMLSpace(c);
```
Do we need to check both `== ' '` and `isHTMLSpace`? Not related to this patch, but something I noticed.

cdumez replied:
I noticed the same thing and my bet is that this is an ordering optimization. I'll A/B test it to see if we can drop.

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -677,19 +675,19 @@ class HTMLFastPathParser {
         }
 
-        CharSpan reference = CharSpan { start, static_cast<size_t>(m_position - start) - 1 };
+        CharSpan reference = CharSpan { start, static_cast<size_t>(m_parsingBuffer.position() - start) - 1 };
```
Could use a briefer syntax without `= CharSpan`.

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -643,32 +639,34 @@ class HTMLFastPathParser {
-        if (Char quoteChar = peekNext(); quoteChar == '"' || quoteChar == '\'') {
-            start = ++m_position;
-            while (m_position != m_end && peekNext() != quoteChar) {
+        const Char* start = m_parsingBuffer.position();
```
Would like to use more `auto` and `auto*` in this code.

cdumez committed:
Use StringParsingBuffer in HTMLFastPathParser
https://bugs.webkit.org/show_bug.cgi?id=252407

Reviewed by Darin Adler.

Use StringParsingBuffer in HTMLFastPathParser. It makes the code clearer and
look safer. This is performance neutral on Speedometer.

* Source/WTF/wtf/text/StringParsingBuffer.h:
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::isQuoteCharacter):
(WebCore::isValidUnquotedAttributeValueChar):
(WebCore::HTMLFastPathParser::HTMLFastPathParser):
(WebCore::HTMLFastPathParser::parseCompleteInput):
(WebCore::HTMLFastPathParser::scanText):
(WebCore::HTMLFastPathParser::scanEscapedText):
(WebCore::HTMLFastPathParser::scanTagName):
(WebCore::HTMLFastPathParser::scanAttributeName):
(WebCore::HTMLFastPathParser::scanAttributeValue):
(WebCore::HTMLFastPathParser::scanEscapedAttributeValue):
(WebCore::HTMLFastPathParser::scanHTMLCharacterReference):
(WebCore::HTMLFastPathParser::parseChildren):
(WebCore::HTMLFastPathParser::parseAttributes):
(WebCore::HTMLFastPathParser::parseContainerElement):
(WebCore::HTMLFastPathParser::isValidUnquotedAttributeValueChar): Deleted.
(WebCore::HTMLFastPathParser::skipWhitespace): Deleted.
(WebCore::HTMLFastPathParser::peekNext): Deleted.
(WebCore::HTMLFastPathParser::consumeNext): Deleted.

Canonical link: https://commits.webkit.org/260417@main
